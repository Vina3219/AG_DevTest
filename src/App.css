.app {
  text-align: center;
  display: flex;
  flex-direction: column;
  height: 100vh;
}

.app-header {
  background-color: #f5f5f5;
  height: 70px;
  display: flex;
  justify-content: space-between;
  padding: 0 24px;
  align-items: center;
}

.app-header__title {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
}

.app-header__title h2 {
  margin: 0;
  font-weight: 500;
  color: #575757;
}

.app-header__title span {
  color: #bebebe;
}

.app-header__select {
  display: flex;
  align-items: flex-end;
  height: 100%;
  margin-bottom: 12px;
}

.app hr {
  margin: 0;
  border-top: 1px solid #bebebe;
  margin-inline-start: 10px;
  margin-inline-end: 10px;
}

.app-footer {
  display: flex;
  height: 70px;
  background-color: #f5f5f5;
  justify-content: flex-end;
  align-items: baseline;
  padding: 8px;
  padding-right: 48px;
}
