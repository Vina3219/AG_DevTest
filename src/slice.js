import { createSlice } from '@reduxjs/toolkit';
import { loadProducts, getTotal } from './services'

const initialState = {
    products: [],
    pageNum: 1,
    productsCountPerPage: 8,
    totalProductsCount: 0
}

const slice = createSlice({
    name: 'app',
    initialState,
    reducers:{
        setPageNum: (state, action) => {
            state.pageNum = action.payload.pageNum;
            state.products = loadProducts(state.pageNum, state.productsCountPerPage);
        },
        setProductsPerPage: (state, action) => {
            state.productsCountPerPage = action.payload.productsCountPerPage;
            state.pageNum = Math.ceil(state.products[0].id/action.payload.productsCountPerPage);
            state.products = loadProducts(state.pageNum, state.productsCountPerPage);
        },
        loadPage: (state) => {
            state.products = loadProducts(state.pageNum, state.productsCountPerPage);
            state.totalProductsCount = getTotal();
        }
    }
})

export const { setPageNum, setProductsPerPage: setProductsCountPerPage, loadPage } = slice.actions;

export const selectProducts = (state) => state.app.products;
export const selectPageNum = (state) => state.app.pageNum;
export const selectTotalPage = (state) => Math.ceil(state.app.totalProductsCount/state.app.productsCountPerPage);
export const selectProductsCountPerPage = (state) => state.app.productsCountPerPage;
export const selectTotalProductsCount = (state) => state.app.totalProductsCount;

export default slice.reducer;
