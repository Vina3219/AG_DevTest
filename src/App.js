import "./App.css";

import { useSelector, useDispatch } from "react-redux";
import { selectTotalProductsCount, loadPage } from "./slice";
import ProductsCountPerPageSelect from "./components/productsCountPerPageSelect/ProductsCountPerPageSelect";
import Pagination from "./components/pagination/Pagination";
import React, { useEffect } from "react";
import ProductsList from "./components/productsList/ProuctsList";

const App = () => {
  const dispatch = useDispatch();
  const totalProuductCount = useSelector(selectTotalProductsCount);

  useEffect(() => {
    dispatch(loadPage());
  }, [dispatch]);

  return (
    <div className="app">
      <header className="app-header">
        <div className="app-header__title">
          <h2>All Products</h2>
          <span>{totalProuductCount} Products</span>
        </div>

        <div className="app-header__select">
          <ProductsCountPerPageSelect />
        </div>
      </header>

      <hr />

      <ProductsList />

      <footer className="app-footer">
        <Pagination />
      </footer>
    </div>
  );
};

export default App;
