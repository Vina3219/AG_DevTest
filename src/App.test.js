import { render } from "@testing-library/react";
import { useDispatch } from "react-redux";
import App from "./App";
import { loadPage } from "./slice";

jest.mock("react-redux", () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn((selector) => selector()),
}));

jest.mock("./slice", () => ({
  ...jest.requireActual("./slice"),
  selectTotalProductsCount: jest.fn(),
  loadPage: jest.fn(),
}));

jest.mock(
  "./components/productsCountPerPageSelect/ProductsCountPerPageSelect",
  () => jest.fn()
);

jest.mock("./components/pagination/Pagination", () => jest.fn());

jest.mock("./components/productsList/ProuctsList", () => jest.fn());

const dispatchSpy = jest.fn();

describe("App", () => {
  beforeEach(() => {
    useDispatch.mockReset().mockReturnValue(dispatchSpy);
    loadPage.mockReset().mockReturnValue("load page");
  });

  it("matches snapshot", () => {
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("loads page", () => {
    render(<App />);
    expect(dispatchSpy).toBeCalledTimes(1);
    expect(dispatchSpy).toBeCalledWith("load page");
  });
});
