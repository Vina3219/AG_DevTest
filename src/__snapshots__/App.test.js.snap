// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="app"
  >
    <header
      class="app-header"
    >
      <div
        class="app-header__title"
      >
        <h2>
          All Products
        </h2>
        <span>
           Products
        </span>
      </div>
      <div
        class="app-header__select"
      />
    </header>
    <hr />
    <footer
      class="app-footer"
    />
  </div>
</DocumentFragment>
`;
